File: framework/src/main/java/com/software/dm/swallow/stormy/algoac/inter/AbstractState.java
package com.software.dm.swallow.stormy.algoac.inter;

import java.util.Collection;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

/**
 * @param <T>
 * @author DM
 */
public abstract class AbstractState<T> implements State<T, AbstractState<T>> {

    protected final int depth;
    protected final AbstractState<T> rootState;
    private AbstractState<T> failure = null;
    private Set<Object> output = null;
    protected Map<T, AbstractState<T>> success = new HashMap<T, AbstractState<T>>();
    protected boolean prepared = false;

    public AbstractState() {
        this(0);
    }

    public AbstractState(int depth) {
        this.depth = depth;
        rootState = this.depth == 0 ? this : null;
    }


    public AbstractState<T> nextState(T c) {
        return this.nextState(c, false);
    }


    public AbstractState<T> nextStateIgnoreRootState(T c) {
        return this.nextState(c, true);
    }

    public AbstractState<T> nextState(T c, boolean ignoreroot) {
        AbstractState<T> currentState = this.success.get(c);
        if (!ignoreroot && currentState == null && this.rootState != null) {
            return this.rootState;
        }
        return currentState;
    }


    public Collection<T> getValues() {
        return this.success.keySet();
    }


    public Collection<AbstractState<T>> getStates() {
        return this.success.values();
    }

    public void addOutput(Object output) {
        if (this.output == null)
            this.output = new HashSet<Object>();
        this.output.add(output);
    }

    public void addAllOutput(Set<Object> output) {
        if (output == null || output.isEmpty())
            return;
        if (this.output == null)
//			return;
            this.output = new HashSet<Object>();
        this.output.addAll(output);
    }

    public Set<Object> getOutput() {
//		return this.output == null ? new HashSet<Object>() : this.output;
        return this.output;
    }

    public AbstractState<T> getFailure() {
        return failure;
    }

    public void setFailure(AbstractState<T> failure) {
        this.failure = failure;
    }

    public int getDepth() {
        return depth;
    }

}

File: framework/src/main/scala/com/software/dm/swallow/stormy/scala/algoac/inter/AbstractState.scala
package com.software.dm.swallow.stormy.scala
package algoac
package inter


import scala.collection.mutable.{HashMap, HashSet, Set}
import scala.collection.Iterable


/**
  * @author darlingming@126.com
  * @param depth
  * @tparam T
  */
abstract class AbstractState[T](val depth: Int) extends State[T, AbstractState[T]] {

  final protected var rootState = if (this.depth == 0) this else null

  private var failure: AbstractState[T] = null
  private var output: Set[Any] = null
  protected var success = new HashMap[T, AbstractState[T]]
  protected var prepared = false


  def nextState(c: T): AbstractState[T] = this.nextState(c, false)

  def nextStateIgnoreRootState(c: T): AbstractState[T] = this.nextState(c, true)

  def nextState(c: T, ignoreroot: Boolean): AbstractState[T] = {
    val currentState = this.success.get(c).getOrElse(null)
    if (!ignoreroot && currentState == null && this.rootState != null) return this.rootState
    currentState
  }

  def getValues: Iterable[T] = this.success.keys

  def getStates: Iterable[AbstractState[T]] = this.success.values

  def addOutput(output: Any): Unit = {
    if (this.output == null) this.output = new HashSet[Any]
    this.output.add(output)
  }

  def addAllOutput(output: Set[Any]): Unit = {
    if (output == null || output.isEmpty) return
    if (this.output == null) { //			return;
      this.output = new HashSet[Any]
    }
    this.output.add(output)
  }

  def getOutput: Set[Any] = { //		return this.output == null ? new HashSet<Object>() : this.output;
    this.output
  }

  def getFailure: AbstractState[T] = failure

  def setFailure(failure: AbstractState[T]): Unit = {
    this.failure = failure
  }

  def getDepth: Int = depth
}
